### PR TITLE
(1263) Make BaseHMPPSClient cache-aware

### PR DESCRIPTION
**Pull request by anandelangovan1**
## Context

[Trello card 1263](https://trello.com/c/hXxe9Tc4/1263-set-up-redis-caching-on-api-referral-summaries-l)

To make our service as efficient as possible, we're going to fetch data from the Prisoner Search and Prison Register APIs in advance and cache the result. This will also give us the ability to sort the data coming back from the paginated API response.

## Changes in this PR

This ticket does the leg work for making our base client cache aware, adding a new `WebClientCache` class to the service with functions for fetching data from a Redis Cache.
 
The `WebClientCache` is injected into the `BaseHMPPSClient` so it can be used on all requests to return a cached value before making a fresh request for data from another source e.g. the Prisoner Search API.

There's currently no caching of data taking place at the moment. This will come in a subsequent PR, but the changes in this commit should allow the client to start making use of this caching functionality once we start fetching storing data from the Prisoner Search and Prison Register services in advance.

This has been heavily inspired (read "copied from") the CAS team's approach [here](https://github.com/ministryofjustice/hmpps-approved-premises-api/blob/main/src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/client/WebClientCache.kt).

## Release checklist

[Release process documentation](../doc/how-to/perform-a-release.md)

As part of our continuous deployment strategy we must ensure that this work is
ready to be released once merged.

### Pre-merge

- [ ] There are changes required to the Accredited Programmes UI for this change to work...
  - [ ] ... and they been released to production already

### Post-merge

- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-preprod-environment) release to preprod
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-production-environment) release to prod

### CODE REVIEW

**Comment by Gweaton** [comment]
Anand's off today, I'm just going to tidy up this branch, looks like it's got some WIP commits from the subsequent branch on it. Will also add a bit more context to commit messages.

File: src/main/resources/application.yml [hunk 1]
```diff
@@ -123,3 +131,6 @@ log-client-credentials-jwt-info: false
 upstream-timeout-ms: 10000
 case-notes-service-upstream-timeout-ms: 30000
 max-response-in-memory-size-bytes: 750000
+
+preemptive-cache-key-prefix: ""
```
Comment (stevomcallister) [inline src/main/resources/application.yml:135]: Does this need a value?

**Review by nsmj63-dxw — APPROVED** [review]
Masterfully done. Simple, easily reversible, and much less complicated than the CAS setup. 👍

File: src/main/resources/application.yml (hunk 1 above)
Comment (anandelangovan1) [reply src/main/resources/application.yml:135]: yes, needed as part of this change. 
I think we plan on adding the prefix for each of the caches we have.

**Pull request merged**